pinot-common/src/main/java/org/apache/pinot/common/upsert/hash/UpsertHashFunctionFactory.java

```diff
@@ -0,0 +1,69 @@
+package org.apache.pinot.common.upsert.hash;
+
+import java.lang.reflect.Constructor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.spi.utils.PinotReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class UpsertHashFunctionFactory {
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertHashFunctionFactory.class);
+  private static final Map<String, Class<UpsertHashFunction>> HASH_FUNCTION_MAP = new HashMap<>();
+  private static final Object INIT_LOCK = new Object();
+  private volatile static boolean _isLoaded = false;
+
+  private UpsertHashFunctionFactory() {
+  }
+
+  public static UpsertHashFunction create(String functionName) {
+    if (!_isLoaded) {
```

Comment:
Ideally we want to load all functions upfront to guarantee performance at ingestion time. It can also simply the locks. See `FunctionRegistry` for example. We may init it in `ServerInstance`